linalg/constraints.cpp

```diff
@@ -0,0 +1,893 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "constraints.hpp"
+
+#include "../fem/pfespace.hpp"
+
+#include <set>
+
+namespace mfem
+{
+
+Eliminator::Eliminator(const SparseMatrix& B, const Array<int>& lagrange_tdofs_,
+                       const Array<int>& primary_tdofs_,
+                       const Array<int>& secondary_tdofs_)
+   :
+   lagrange_tdofs(lagrange_tdofs_),
+   primary_tdofs(primary_tdofs_),
+   secondary_tdofs(secondary_tdofs_)
+{
+   MFEM_VERIFY(lagrange_tdofs.Size() == secondary_tdofs.Size(),
+               "Dof sizes don't match!");
+
+   Bp.SetSize(lagrange_tdofs.Size(), primary_tdofs.Size());
+   B.GetSubMatrix(lagrange_tdofs, primary_tdofs, Bp);
+
+   Bs.SetSize(lagrange_tdofs.Size(), secondary_tdofs.Size());
+   B.GetSubMatrix(lagrange_tdofs, secondary_tdofs, Bs);
+   BsT.Transpose(Bs);
+
+   ipiv.SetSize(Bs.Height());
+   Bsinverse.data = Bs.GetData();
```

Comment:
In order to reduce potential issues for those using GPU code with these files, I would suggest for the time changing (where possible) all of these `GetData` in this file to either `HostRead()`, `HostReadWrite()`, or `HostWrite()` just so we know for sure the data is on the host and in-sync with everything. It should also make things easier on later on when trying to convert this over to run on the GPU.